File: UnitTests/CoffeeMaker/CSHARP/CoffeeMaker.Test/UnitTests/CoffeeTests.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using CoffeeMaker.interfaces;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using NUnit.Framework;
using Assert = Microsoft.VisualStudio.TestTools.UnitTesting.Assert;
using CollectionAssert = Microsoft.VisualStudio.TestTools.UnitTesting.CollectionAssert;
using IgnoreAttribute = Microsoft.VisualStudio.TestTools.UnitTesting.IgnoreAttribute;

namespace CoffeeMaker.Test.IntegrationTests
{
    [TestClass]
    public class CoffeeSelectTests
    {

        private ICoffee wetBeanCoffee;

        [TestInitialize]
        public void Init() {

            wetBeanCoffee = new WetBeansCoffee();
        }

        [TestCleanup]
        public void Cleanup()
        {
            wetBeanCoffee = null;
        }

       [Ignore("Do not run this test as we are not ready for it.")]
       [TestMethod]
        public void DryingBeansForConsistency()
        {

            string consistencyValue = wetBeanCoffee.dry_beans();

            for (int i = 0; i < 5; i++)
            {
                string test = wetBeanCoffee.dry_beans();

                if (test != consistencyValue)
                {
                    Assert.Inconclusive("the value of dry beans is inconsistent");
                }
            }
        }

        [Repeat(11)]
        [TestMethod]
        public void DryingBeansForConsistancyWithException()
        {

            string consistencyValue = wetBeanCoffee.dry_beans();
            string test = wetBeanCoffee.dry_beans();

            if (test != consistencyValue)
            {
                Assert.Inconclusive("the value of dry beans is inconsistent");
            }
            
        }

        [Order(1)]
        [TestMethod]
        public void TestDryingBeansHasValue()
        {
            //value is different everytime
            string consistencyValue = wetBeanCoffee.dry_beans();
            Assert.IsFalse(string.IsNullOrEmpty(consistencyValue));
        }

        [Order(2)]
        [TestMethod]
        public void TestBeanVolume()
        {
            int beanVolume = wetBeanCoffee.GetBeans();
            Assert.AreEqual(beanVolume, 4); 
        }

        [TestMethod]
        public void TestLabelsAreEqual()
        {
            CoffeeModel coffee = new CoffeeModel();
            coffee = wetBeanCoffee.AssignCoffeeProperties(1, "Frapachino", true, "France");

            Assert.AreEqual<string>(coffee.Label, "Frapachino");
        }

        [TestMethod]
        public void TestIfTwoCoffeeObjectsAreTheSame()
        {

            CoffeeModel coffeeOne = new CoffeeModel();
            CoffeeModel coffeeTwo = new CoffeeModel();
            coffeeOne = wetBeanCoffee.AssignCoffeeProperties(1, "Frapachino", true, "France");
            coffeeTwo = wetBeanCoffee.AssignCoffeeProperties(1, "Frapachino", true, "France");

            Assert.AreNotSame(coffeeOne, coffeeTwo);
            // watch this gotcha.
            Assert.AreSame(coffeeOne.Label, coffeeTwo.Label);

        }

        [TestMethod]
        public void TestCoffeeGroup()
        {
            CoffeeModel coffeeOne = new CoffeeModel();
            coffeeOne = wetBeanCoffee.AssignCoffeeProperties(1, "Frapachino", true, "France");
            List<CoffeeModel> OrderredByGroup = wetBeanCoffee.GroupOrder();

            CollectionAssert.AllItemsAreUnique(OrderredByGroup);

            //note it is the same type with same values but it is a different object
            CollectionAssert.DoesNotContain(OrderredByGroup, coffeeOne);
            CollectionAssert.AllItemsAreNotNull(OrderredByGroup);
        }

        [TestMethod()]
        [ExpectedException(typeof(ArgumentNullException), "To throw an exception when argument is null.")]
        public void WeakCoffeeGrindBeansTestWithInvalidArgument()
        {
            wetBeanCoffee.GrindBeans(null);
        }
    }
}

File: UnitTests/CoffeeMaker/Java/src/test/java/CoffeeTests.java
import Interfaces.*;
import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.jupiter.api.Assertions.*;
import Models.CoffeeModel;
import org.junit.jupiter.api.*;

import java.io.IOException;
import java.util.List;

public class CoffeeTests {

    private static ICoffee wetBeanCoffee;

    @BeforeAll
    public static void init() {
        wetBeanCoffee = new WetBeansCoffee();
    }

    @BeforeEach
    public void initBeforeEachTest() {
        System.out.println("Setting it up before every test.");
    }

    @AfterEach
    public void cleanupAfterEachTest()
    {
        System.out.println("Clean up after every test");
    }

    @AfterAll
    public static void cleanup()
    {
        wetBeanCoffee = null;
    }

    @Disabled("Do not run this test as we are not ready for it.")
    @Test
    public void testDryingBeansForConsistency() {

        String consistencyValue = wetBeanCoffee.dry_beans();
        String test = wetBeanCoffee.dry_beans();
        if (test != consistencyValue)
        {
            Assertions.fail("the value of dry beans is inconsistent");
        }
    }

    @RepeatedTest(11)
    @Test()
    public void testDryingBeansForConsistancyWithException() {

        String consistencyValue = wetBeanCoffee.dry_beans();
        String test = wetBeanCoffee.dry_beans();
        if (test != consistencyValue)
        {
            System.out.println("the value of dry beans is inconsistent");
        }
    }

    @Test
    @Order(1)
    public void testDryingBeansHasValue()
    {
        String consistencyValue = wetBeanCoffee.dry_beans();
        assertNotEquals(consistencyValue, "");
    }

    @Test
    @Order(2)
    public void testBeanVolume()
    {
        int beanVolume = wetBeanCoffee.getBeans();
        assertEquals(beanVolume, 4);
    }

    @Test
    public void testLabelsAreEqual() {
        CoffeeModel coffee = new CoffeeModel();
        coffee = wetBeanCoffee.assignCoffeeProperties(1, "Frapachino", true, "France");
        assertEquals(coffee.getLabel(), "Frapachino");
    }

    @Test
    public void testComparingObjects()
    {
        CoffeeModel coffeeOne;
        CoffeeModel coffeeTwo;

        coffeeOne = wetBeanCoffee.assignCoffeeProperties(1, "Frapachino", true, "France");
        coffeeTwo = wetBeanCoffee.assignCoffeeProperties(1, "Frapachino", true, "France");

        CoffeeModel coffeeThree = coffeeOne;

        //are the same as the point to the same object
        assertSame(coffeeOne,coffeeThree);
        assertEquals(coffeeOne,coffeeThree);

        // are not the same although same type and same values
        assertNotSame(coffeeOne, coffeeTwo);
        assertNotEquals(coffeeOne, coffeeTwo);

        // watch this gotcha.  This should return false as they are not the same item, they do have the same value and will fail if they have different values.
        assertSame(coffeeOne.getLabel(), coffeeTwo.getLabel());
        // should use below for consistency
        assertEquals(coffeeOne.getLabel(), coffeeTwo.getLabel());
    }

    @Test
    public void testCoffeeGroup()
    {
        CoffeeModel coffeeOne;
        coffeeOne = wetBeanCoffee.assignCoffeeProperties(1, "Frapachino", true, "France");
        List<CoffeeModel> OrderredByGroup = wetBeanCoffee.groupOrder();

        assertThat(OrderredByGroup, not(hasItem(coffeeOne)));
        assertThat(OrderredByGroup, not(nullValue()));
    }

    @Test
    public void testWeakCoffeeGrindBeansTestWithInvalidArgument()
    {
        assertThrows(NullPointerException.class, () -> {wetBeanCoffee.grindBeans(null);});
    }

    @Test
    public void getGranniesSecretReceipt() throws IOException {
        assertFalse(wetBeanCoffee.runGetCoffee());
    }

}
